source/jquery.flot.drawSeries.js

```diff
@@ -23,124 +23,154 @@ This plugin is used by flot for drawing lines, plots, bars or area.
                 my = null,
                 i = 0;
 
-            ctx.beginPath();
-            for (i = ps; i < points.length; i += ps) {
+            var initPoints = function () {
                 x1 = points[i - ps];
                 y1 = points[i - ps + 1];
                 x2 = points[i];
                 y2 = points[i + 1];
+            };
 
-                if (x1 === null || x2 === null) {
+            var handleSteps = function () {
+                if (mx !== null && my !== null) {
+                    // if middle point exists, transfer p2 -> p1 and p1 -> mp
+                    x2 = x1;
+                    y2 = y1;
+                    x1 = mx;
+                    y1 = my;
+
+                    // 'remove' middle point
                     mx = null;
                     my = null;
-                    continue;
-                }
-
-                if (isNaN(x1) || isNaN(x2) || isNaN(y1) || isNaN(y2)) {
-                    prevx = null;
-                    prevy = null;
-                    continue;
-                }
-
-                if (steps) {
-                    if (mx !== null && my !== null) {
-                        // if middle point exists, transfer p2 -> p1 and p1 -> mp
-                        x2 = x1;
-                        y2 = y1;
-                        x1 = mx;
-                        y1 = my;
 
-                        // 'remove' middle point
-                        mx = null;
-                        my = null;
-
-                        // subtract pointsize from i to have current point p1 handled again
-                        i -= ps;
-                    } else if (y1 !== y2 && x1 !== x2) {
-                        // create a middle point
-                        y2 = y1;
-                        mx = x2;
-                        my = y1;
-                    }
+                    // subtract pointsize from i to have current point p1 handled again
+                    i -= ps;
```

Comment:
I think if we're going to reorganize this code in such a way, I would prefer that we not hide the mutation of `i` within this function. Additionally, I think it would be better if we made any function that operated on the current value of `i` to have `i` passed into via an argument (i.e. `initPoints` now takes `i` as a parameter).

This will make the final invocation of `initPoints` clearer for what it is operating on, instead of it relying on whatever state `i` has when exiting the `for` loop.

Reply:
Thank you very much for your quick response and feedback. 
I think it may be better the way you suggested. I will do that asap.

Otherwise, I was thinking about one more point; what do you think about the fact that after the loop, I do not do the "continue logic" anymore. I am not sure that is needed, but If I want to keep it exactly as it was previously in the loop, maybe I have to check if it should exit from that if block, or execute further...

Reply:
Sorry, I guess I'm not following exactly what you are referring to. In the previous code, I do not see any "continue logic" occurring after loop.

Reply:
I was referring to "continue" statements which were previously in the loop; 
you see that after my refactoring, I'm still keeping the same login in the loop, 
but after the loop I try to do quite the same (in order to have like "one more iteration", to properly connect the last points), BUT this time I do not have the returns going out of the if block in case the "continue" is triggered in any functions.

I hope now it's clear what I mean.

Reply:
Ah I see. Yeah, what's interesting here is that in the previous implementation, if any `continue` was tripped than we wouldn't ultimately call `moveTo` or `lineTo` in that iteration. In the new implementation, I suppose it's conceivable that one of the clipping functions would return `true` (same as `continue`), and we will still not only call any successive clipping functions, but also before a `lineTo`. I'm not sure what the consequences of that are.

At the moment, it's hard for me to wrap my head around what the behavior will ultimately be. I think I would need to try and create each of the scenarios (or at least one of them) to see what happens (this code is _old_ and I didn't write it :P).

Reply:
Yes, that's it. So in any case I think I can write the logic to keep it consistent with what it was previously, this way we can avoid some strange scenarios as you pointed out with `moveTo` or `lineTo`.

I will commit in few minutes.

Reply:
I've tested it in my project and it works very well. 
Let me know if these changes look fine for you or if there is something else I can do. 

Thank you very much for your time. 🙏